designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,585 @@
+1. [specifying ignore patterns in the config](https://github.com/eslint/eslint/issues/10891)
+1. [specifying `--ext` information in the config](https://github.com/eslint/eslint/issues/11223)
+1. [using `extends` in `overrides`](https://github.com/eslint/eslint/issues/8813) 
+1. [Customize merging of config options](https://github.com/eslint/eslint/issues/9192)
+
+
+The only reason that these requests are difficult to implement in ESLint is because of how complex the `.eslintrc` configuration format is. Any changes made to any part of `.eslintrc` processing end up affecting millions of ESLint installations, so we have ended up stuck. The complicated parts of `.eslintrc` include:
+
+1. Resolution behavior of modules (`plugins`, `parser`, `extends`)
+1. Merging of cascading config files in a directory structure
+1. Merging of config files via `extends`
+1. Overriding of configuration via `overrides`
+
+All of these complexities arise from `.eslintrc` format of describing what should happen rather than how it should happen. We keep trying to anticipate how users want things to happen and guessing wrong has led to increased complexity.
+
+This RFC proposes a fresh start for configuration in ESLint that takes into account all of the requests we've received over the years. By simplifying and stripping configuration we can actually make a configuration system that is flexible enough to accomodate changes in the future and easy enough to implement that we won't be afraid to do so.
+
+## Detailed Design
+
+Design Summary:
+
+1. Introduce a new `eslint.config.js` configuration file format
+1. Searching for `eslint.config.js` starts from the current working directory and continues up
+1. All `eslint.config.js` files are treated as if they have `root: true`
+1. There is no automatic merging of config files
+1. The `eslint.config.js` configuration is not serializable and objects such as functions and plugins may be added directly into configuration
+1. Remove the concept of environments (`env`)
+1. Remove `.eslintignore` and `--ignore-file` (no longer necessary)
+1. Remove `--rulesdir`
+1. Create a `@eslint/config` package to help users merge configs
+
+### The `eslint.config.js` File
+
+The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports a `config` object: 
+
+```js
+exports.config = {
+    files: ["*.js"],
+    ignore: ["*.test.js"],
+    extends: [],
+    globals: {},
+    settings: {},
+    processor: object,
+    parser: object,
+    parserOptions: {},
+    ruledefs: {},
+    rules: {}
+};
+```
+
+The following keys are new to the `eslint.config.js` format:
+
+* `files` - **Required.** Determines the glob file patterns that this configuration applies to.
+* `ignore` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`.
```

Comment:
the common case here is "gitignore", or "gitignore plus a few extras" - it'd be great if there was a way to let eslint pick this up automatically without having to import and split its contents here.

Reply:
I think letting users manually read in whatever file they want and process it however they want is the best approach here. It's easy to think that everyone follows a particular pattern, but I don't think we want to canonize any approach that involves files that ESLint doesn't own directly. (Perhaps there's room for a helper utility here, though I might want to wait for the first group of feedback from real users to decide what that looks like.)

Reply:
This is a very small corner of this proposal, so I'd suggest we table this discussion for now and revisit when the first implementation and batch of feedback has been collected to see if there is any complaints about this. For now, I think keeping this process manual sticks with the overall goals of this design.

Reply:
As long as the possibility remains to add "use gitignore" in an ergonomic fashion, that seems totally reasonable.